csrc/device_lower/pass/double_buffer.cpp

```diff
@@ -419,6 +729,191 @@ class DoubleBufferLoopNestInspector : private kir::IrVisitor {
   InsertionInfo insertion_info_;
 };
 
+// Creates pre-prologue pieces of code needed for proper handling
+//  async TMA memory copies - moving allocation outside of main loop
+//  and initialization of mbarriers.
+//
+// Expected result:
+//   mbarrier and token smem allocations
+//   if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
+//     for (unsigned i = 0; i < stages; ++i) {
+//       mbarrier::init(...);
+//     }
+//   }
+class CpAsyncBulkPrePrologue : public kir::IrVisitor {
```

Comment:
Should we add a `DoubleBufferLoopStage::PrePrologue` and a `DoubleBufferLoopStage::PostEpilogue` and move the code below to `DoubleBufferLoopCloner`?

Reply:
That's a good point.

For now I have it as separate class of functional classes - the main reason is that `DoubleBufferLoopCloner` returns `kir::ForLoop` while pre-prologue/post-epilogue are a simple collection (a list) of expressions. But you are right, we could have this done in smarter way where these two new 'regions' are handled along prolog/main/epilogue. I will try to unify this part of code, or at least leave a comment that architecture of this part of system should be reconsidered.

Thanks!